src/irmin/proof.ml

```diff
@@ -288,12 +306,19 @@ struct
         pp_hash h
 
   let check_node_integrity v h =
-    let h' = Node_hash.hash v in
+    (* Using [N..hash] because using [Irmin.Hash.Typed] here would
+       not be fine as it may trigger IOs. *)
```

Comment:
What do we mean by "trigger IOs" here? I thought `Irmin.Hash.Typed` was purely in-memory computation (no backend interaction involved).

Reply:
`Irmin.Hash.Typed` calls `Repr.pre_hash` on inodes, which may need to load more data. See https://github.com/mirage/irmin/blob/main/src/irmin-pack/inode.ml#L1842

Reply:
We could maybe change the new `val hash : t -> hash` to `val cached_hash : t -> hash option` or `val hash_exn : ?force:bool -> t -> hash`

Reply:
I think that this point was also the reason why we failed to migrate the inodes to irmin core. We were unable to abstract over IO the inode typerepr because of the pre_hash

Reply:
See #1323

Reply:
Ah, interesting. I hadn't really processed that pre-hash must work in this way over inodes, but it makes sense.

I think `val hash : t  -> hash` makes sense as an API, even if it clashes a bit with how we are doing things elsewhere, but I have no strong favourite of the options you suggest.

Reply:
I think I will choose something else than `val hash : t -> hash` because I want to explicitly forbid IOs during that hash computation.

Since we already have `?cache:bool` parameters in the node module type, I think I will go with `val hash_exn : ?force:bool -> t -> hash`.